### ISSUE TITLE
@dataclass + Subscripted Generics error

### ISSUE THREAD

**Issue by elbaro** [astral-sh/ty#853]
### Summary

```py
from dataclasses import dataclass


@dataclass
class Wrap[T]:
    inner: T


class Object: ...


@dataclass
class ObjectWrap(Wrap[Object]): ...


w = ObjectWrap(inner=Object())
print(w)
```

```
  --> tt.py:16:16
   |
16 | w = ObjectWrap(inner=Object())
   |                ^^^^^^^^^^^^^^ Expected `T`, found `Object`
17 | print(w)
   |
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```

If you comment `@dataclass` on the outer class, it passes:
```py
# @dataclass
class ObjectWrap(Wrap[Object]): ...
```

### Version

ty 0.0.1-alpha.15

**Comment by carljm** [comment]
This [desugared version](https://play.ty.dev/89725d2b-221b-4e10-92ad-dbab2a1463dd) doesn't show the error, so it seems this is dataclasses-specific:

```py
class Wrap[T]:
    inner: T

    def __init__(self, inner: T) -> None:
        self.inner = inner

class Object: ...

class ObjectWrap(Wrap[Object]): ...

w = ObjectWrap(inner=Object())
print(w)
```

**Comment by carljm** [comment]
I wonder if `apply_type_mapping` isn't descending into `Callable` type attributes (so we fail to properly specialize the synthetic inherited `__init__` method)

**Comment by sharkdp** [comment]
> I wonder if `apply_type_mapping` isn't descending into `Callable` type attributes (so we fail to properly specialize the synthetic inherited `__init__` method)

It was a slightly different bug where the specialization was not passed down to generic dataclass *base classes* (generic dataclasses were already working correctly). I pushed a PR with a fix here: https://github.com/astral-sh/ruff/pull/19472
